### PR TITLE
Cache zeroconf device info

### PR DESCRIPTION
**Pull request by theychx**
Howdy
For some time now, Google devices have stopped providing manufacturer/model_name info through DIAL, breaking our device type detection stuff in `StreamInfo`. So I've implemented a change to our caching, so we cache all the info we need from zeroconf. This work also turned out make some code a bit leaner.

PS: Your changelog thingy is indeed out of control :smile:

### CODE REVIEW

File: catt/controllers.py [hunk 1]
```diff
@@ -215,32 +219,33 @@ def __init__(self):
...
+            return None
         if name:
             fetched = data.get(name)
         else:
             # When the user does not specify a device, we need to make an attempt
             # to consistently return the same IP, thus the alphabetical sorting.
             fetched = data[min(data, key=str)]
-        return (fetched["ip"], fetched.get("group_port", 0)) if fetched else (None, None)
+        return (
+            CCInfo(fetched["ip"], fetched["port"], fetched["manufacturer"], fetched["model_name"], fetched["cast_type"])
```
Comment (theychx) [inline catt/controllers.py]: Is there a more elegant way of doing this?

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,8 @@
 
 ### Fixes
 
+* Add socket based fallback approach in get_local_ip. [Stavros Korokithakis]
```
Comment (skorokithakis) [inline CHANGELOG.md]: Oh geez.

File: catt/stream_info.py [hunk 3]
```diff
@@ -7,27 +7,27 @@
 from .util import get_local_ip, guess_mime
 
 AUDIO_DEVICE_TYPES = ["audio", "group"]
-ULTRA_MODELS = [("Xiaomi", "MIBOX3"), ("Google Inc.", "Chromecast Ultra")]
+ULTRA_MODELS = [("Xiaomi", "MIBOX3"), ("Unknown manufacturer", "Chromecast Ultra")]
```
Comment (theychx) [inline catt/stream_info.py]: The **manufacturer** field is only available though DIAL, so this is now completely unavailable for Google devices.

File: catt/controllers.py (hunk 1 above)
Comment (skorokithakis) [reply catt/controllers.py]: Yep, `CCInfo(**fetched)`.

**Comment by theychx** [comment]
@skorokithakis 
Just FYI, it turns out that those missing info fields are indeed still available via DIAL. It's just not where `pychromecast` is currently looking.

The info is still available here: `http://CAST_IP:8008/ssdp/device-desc.xml`
`pychromecast` is currently using `http://CAST_IP:8008/setup/eureka_info?options=detail`

The problem with the former is that some devices (my cc audio included) respond with a 404, so the `pychromecast` folks actually [switched](https://github.com/balloob/pychromecast/pull/197) from the former to the latter some time ago.

So in conclusion, I still think we should go ahead with this PR.

**Comment by skorokithakis** [comment]
That makes sense, although I wonder why they didn't just fall back to the latter of the former failed. Do you want to address the keyword arguments line first or should I approve now?

File: catt/controllers.py (hunk 1 above)
Comment (theychx) [reply catt/controllers.py]: How can I be sure that the params are fed to `CCInfo` in the right order?

**Comment by theychx** [comment]
> although I wonder why they didn't just fall back to the latter of the former failed

At the time of the PR, it seems like most devices 404'ed on `device-desc.xml`. And besides, `eureka_info` did at that point contain all the info needed.

File: catt/controllers.py (hunk 1 above)
Comment (skorokithakis) [reply catt/controllers.py]: They're passed as keyword parameters, so order doesn't matter.

Comment (theychx) [reply catt/controllers.py]: Cheers. The relation between positional and keyword parameters is confusing in python. Any best practice you can recommend?

Comment (skorokithakis) [reply catt/controllers.py]: Bah I pressed resolve on this by accident, the mobile interface is confusing. Hmm, I tend to call things positionally for small numbers of arguments, if the number of arguments gets larger or they're more convoluted, I tend to specify keywords.

In this case, it's just easier to expand into keywords, rather than specify the parameters one by one.

**Comment by theychx** [comment]
@skorokithakis Did you have any further comments?

File: catt/controllers.py [hunk 4]
```diff
@@ -215,32 +219,29 @@ def __init__(self):
...
             fetched = data.get(name)
         else:
             # When the user does not specify a device, we need to make an attempt
             # to consistently return the same IP, thus the alphabetical sorting.
             fetched = data[min(data, key=str)]
-        return (fetched["ip"], fetched.get("group_port", 0)) if fetched else (None, None)
+        return CCInfo(**fetched) if fetched else None
 
-    def set_data(self, name: str, ip: str, port: int) -> None:  # type: ignore
+    def set_data(self, name: str, device_entry: CCInfo) -> None:  # type: ignore
```
Comment (skorokithakis) [inline catt/controllers.py]: What's the error here that made us ignore the type?

File: catt/stream_info.py (hunk 3 above)
Comment (skorokithakis) [reply catt/stream_info.py]: Should this be a comment in the code?

Comment (theychx) [reply catt/stream_info.py]: It was FYI, but a comment won't hurt. :smiley:

File: catt/controllers.py (hunk 4 above)
Comment (theychx) [reply catt/controllers.py]: Removing the ignore results in this error msg from `mypy`:
```
Signature of "set_data" incompatible with supertype "CattStore"
```
..which still doesn't make sense to me.
